### ISSUE TITLE
jinja templater: Defining and using a macro in the same file causes runtime error

### ISSUE THREAD

**Issue by barrywhart** [sqlfluff/sqlfluff#2456]
## Expected Behaviour
Successful linting
## Observed Behaviour
Runtime error:
```
(sqlfluff-3.10.1) ➜  sqlfluff git:(main) ✗ sqlfluff lint test.sql
WARNING    Unable to lint test.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!
To hide this warning, add the failing file to .sqlfluffignore
Traceback (most recent call last):
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/linter/runner.py", line 103, in run
    yield partial()
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/linter/linter.py", line 598, in lint_rendered
    parsed = cls.parse_rendered(rendered)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/linter/linter.py", line 352, in parse_rendered
    tokens, lvs, config = cls._lex_templated_file(
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/linter/linter.py", line 139, in _lex_templated_file
    tokens, lex_vs = lexer.lex(templated_file)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/parser/lexer.py", line 321, in lex
    segments: Tuple[RawSegment, ...] = self.elements_to_segments(
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/parser/lexer.py", line 348, in elements_to_segments
    source_slice = templated_file.templated_slice_to_source_slice(
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/templaters/base.py", line 294, in templated_slice_to_source_slice
    ts_stop_sf_start, ts_stop_sf_stop = self._find_slice_indices_of_templated_pos(
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/templaters/base.py", line 180, in _find_slice_indices_of_templated_pos
    raise ValueError("Position Not Found")
ValueError: Position Not Found
```
## Steps to Reproduce
Lint this file:
```
{% macro get_keyed_nulls(columns) %}

  COALESCE({{columns}}, MD5(-1))

{% endmacro %}

SELECT
    c1,
    {{ get_keyed_nulls("other_id") }} AS other_id
FROM my_table
```
## Configuration
```
Include your SQLFluff configuration here
```

**Comment by barrywhart** [comment]
DIscovered when investigating #2270.
